src/engine/DebugOverlays.cs

```diff
@@ -0,0 +1,164 @@
+﻿using Godot;
+
+/// <summary>
+///   Main script for debugging.
+///   Partial class: Override functions, debug panel
+/// </summary>
+public partial class DebugOverlays : Control
+{
+    [Export]
+    public NodePath FPSCheckBoxPath = null!;
+
+    [Export]
+    public NodePath PerformanceMetricsCheckBoxPath = null!;
+
+    [Export]
+    public NodePath DebugPanelDialogPath = null!;
+
+    [Export]
+    public NodePath FPSCounterPath = null!;
+
+    [Export]
+    public NodePath PerformanceMetricsPath = null!;
+
+    [Export]
+    public NodePath EntityLabelsPath = null!;
+
+    private static DebugOverlays? instance;
+
+    private CustomDialog debugPanelDialog = null!;
+    private CustomCheckBox fpsCheckBox = null!;
+    private CustomCheckBox performanceMetricsCheckBox = null!;
+    private Control fpsCounter = null!;
+    private CustomDialog performanceMetrics = null!;
+    private Control labelsLayer = null!;
+
+    private DebugOverlays()
+    {
+        instance = this;
+    }
+
+    public static DebugOverlays Instance => instance ?? throw new InstanceNotLoadedYetException();
+
+    public override void _EnterTree()
+    {
+        base._EnterTree();
+
+        Show();
+        InputManager.RegisterReceiver(this);
+
+        // Entity label
+        var rootTree = GetTree();
+        rootTree.Connect("node_added", this, nameof(OnNodeAdded));
+        rootTree.Connect("node_removed", this, nameof(OnNodeRemoved));
+    }
+
+    public override void _ExitTree()
+    {
+        InputManager.UnregisterReceiver(this);
+
+        // Entity label
+        var rootTree = GetTree();
+        rootTree.Disconnect("node_added", this, nameof(OnNodeAdded));
+        rootTree.Disconnect("node_removed", this, nameof(OnNodeRemoved));
+
+        base._ExitTree();
```

Comment:
Base method calls need to be at the start of a method unless specifically that causes a problem.